File: Design-Patterns-Using-Java/Prototype/PrototypePatternShape.java
import java.util.HashMap;
import java.util.Map;

abstract class Shape implements Cloneable {
    private String color;

    public Shape() {
        this.color = "";
    }

    @Override
    public abstract String toString();

    @Override
    protected Object clone() throws CloneNotSupportedException {
        return super.clone();
    }

    public abstract Shape cloneShape();
}

class Rectangle extends Shape {
    @Override
    public String toString() {
        return "Rectangle.";
    }

    @Override
    public Shape cloneShape() {
        try {
            return (Shape) this.clone();
        } catch (CloneNotSupportedException e) {
            e.printStackTrace();
            return null;
        }
    }
}

class Circle extends Shape {
    @Override
    public String toString() {
        return "Circle.";
    }

    @Override
    public Shape cloneShape() {
        try {
            return (Shape) this.clone();
        } catch (CloneNotSupportedException e) {
            e.printStackTrace();
            return null;
        }
    }
}

class ShapeRegistry {
    private static final Map<String, Shape> shapes = new HashMap<>();

    static void addShape(String key, Shape value) {
        shapes.put(key, value);
    }

    static Shape getShape(String key) {
        if (shapes.containsKey(key)) {
            return shapes.get(key).cloneShape();
        }
        return null;
    }

    static void load() {
        addShape("circle", new Circle());
        addShape("rectangle", new Rectangle());
    }
}

public class PrototypePatternShape {
    public static void main(String[] args) {
        ShapeRegistry.load();
        Shape c = ShapeRegistry.getShape("circle");
        Shape r = ShapeRegistry.getShape("rectangle");
        System.out.println(c);
        System.out.println(r);
    }
}

/*
 Circle. Rectangle.
 */

File: Design-Patterns-Using-Kotlin/Prototype/PrototypePatternShape.kt
import java.util.HashMap

abstract class Shape : Cloneable {
    var color: String = ""

    abstract override fun toString(): String

    public override fun clone(): Shape{
        return super.clone() as Shape
    }
}

class Rectangle : Shape() {
    public override fun clone(): Shape {
        return super.clone() as Rectangle
    }

    override fun toString(): String {
        return "Rectangle."
    }
}

class Circle : Shape() {
    public override fun clone(): Shape {
        return super.clone() as Circle
    }

    override fun toString(): String {
        return "Circle."
    }
}

object ShapeRegistry {
    private val shapes: MutableMap<String, Shape> = HashMap()

    fun addShape(key: String, value: Shape) {
        if (!shapes.containsKey(key)) {
            shapes[key] = value
        }
    }

    fun getShape(key: String): Shape {
        return shapes[key]?.clone() ?: throw NoSuchElementException("Shape with key $key not found")
    }

    fun load() {
        addShape("circle", Circle())
        addShape("rectangle", Rectangle())
    }
}

// Client code
fun main() {
    ShapeRegistry.load()
    val c = ShapeRegistry.getShape("circle")
    val r = ShapeRegistry.getShape("rectangle")
    println("$c")
    println("$r")
}

/*
Circle. 
Rectangle.
*/